cacheable.decorator.ts

```diff
@@ -43,7 +31,23 @@ export function Cacheable(cacheConfig: IObservableCacheConfig = {}) {
 
       /* use function instead of an arrow function to keep context of invocation */
       (propertyDescriptor.value as any) = function(...parameters: Array<any>) {
-        const cachePairs: Array<ICachePair<Observable<any>>> = storageStrategy.getAll(cacheKey);
+        this.__cacheId == this.__cacheId || targetCounter++;
```

Comment:
To be honest, I don't like the idea of using `this` in such a functional implementation. What does it point to here?
Also, I'd refrain from maintaining an inline counter here since it seems somehow flaky. Have you considered an implementation around my comment [here](https://github.com/angelnikolov/ngx-cacheable/issues/85#issuecomment-659401755)?
It gives us more power with the context callback since you could provide an arbitrary context which will fit other cases too.

Reply:
`this` is the instance the method is defined on. And you already use it: https://github.com/angelnikolov/ngx-cacheable/blob/27c3433ee6d132d6f798ed00b29e7e63487205ae/cacheable.decorator.ts#L85

> Also, I'd refrain from maintaining an inline counter here since it seems somehow flaky.

why should this be flaky? Another option could be to store the cache on the instance itself. 

> Have you considered an implementation around my comment here?

Yes, but I think it has one flaw: the developer has to think about it, if he doesn't think about it the developer would have the same problem we had with angular-universal.

> It gives us more power with the context callback since you could provide an arbitrary context which will fit other cases too.

yes, but it needs to be explicit, may this could be an additional option. My wish would be if it is easy to use as possible.

> Thank you very much for the effort!
You're very welcome, thanks for your time for commenting on my ideas.

Reply:
The developer wouldn't have to think about it if we provide it as a strategy. For example we can provide a default `UniqueCacheKeyStrategy` which would have your implementation in it so it generates a unique key on every invocation. But the way we hook it to the cache layer should be configurable I think. Different people might have different needs. Let me know what you think.

Reply:
that's sounds good, I will look into it, as soon as possible, thanks for your feedback